source/frontend/editors/RarcEditor.cpp

```diff
@@ -0,0 +1,1035 @@
+  RSL_DEFER(editor->ApplyUpdates());
+
+  constexpr auto dir_flags = ImGuiTreeNodeFlags_OpenOnArrow |
+                             ImGuiTreeNodeFlags_OpenOnDoubleClick |
+                             ImGuiTreeNodeFlags_SpanFullWidth;
+
+  constexpr auto file_flags =
+      ImGuiTreeNodeFlags_Leaf | ImGuiTreeNodeFlags_SpanFullWidth;
+
+  constexpr auto root_flags = dir_flags | ImGuiTreeNodeFlags_DefaultOpen;
+
+  if (m_context_modal_state == ModalState::M_CLOSED && m_operation_modal_state == ModalState::M_CLOSED)
+    m_focused_node = std::nullopt;
+
+  ImGui::Text("Search");
+  ImGui::SameLine();
+  m_model_filter.Draw();
+
+  if (ImGui::BeginTable("TABLE", 3,
+                        ImGuiTableFlags_RowBg |
+                            ImGuiTableFlags_NoSavedSettings)) {
+    RSL_DEFER(ImGui::EndTable());
+
+    ImGui::TableSetupScrollFreeze(3, 1);
+    ImGui::TableSetupColumn("Node Tree");
+    ImGui::TableSetupColumn("Flags", ImGuiTableColumnFlags_WidthFixed, 240.0f);
+    ImGui::TableSetupColumn("Size", ImGuiTableColumnFlags_WidthFixed, 50.0f);
+    ImGui::TableHeadersRow();
+
+    std::vector<std::size_t> walk_stack;
+
+    std::size_t row = 0;
+    for (auto node = rarc.nodes.begin(); node != rarc.nodes.end(); row++) {
+      // We don't render . and ..
+      if (node->is_special_path()) {
+        node++;
+        continue;
+      }
+
+      // Check for ended folders
+      for (auto walk = walk_stack.begin(); walk != walk_stack.end();) {
+        if (*walk == std::distance(rarc.nodes.begin(), node)) {
+          // We assume rightfully that all future walks are also ended
+          std::size_t ended_walks = std::distance(walk, walk_stack.end());
+          for (int i = 0; i < ended_walks; i++) {
+            walk_stack.pop_back();
+            ImGui::TreePop();
+          }
+          break;
+        }
+        walk++;
+      }
+
+      if (!m_model_filter.PassFilter(node->name.c_str())) {
+        node++;
+        continue;
+      }
+
+      ExtensionInfo extension_info;
+      {
+        std::string extension = node->name.substr(node->name.find(".") + 1);
+        if (node->is_folder()) {
+          extension_info = {"Folder", (const char*)ICON_FA_FOLDER};
+        } else if (s_extension_map.contains(extension)) {
+          extension_info = s_extension_map[extension];
+        } else {
+          extension_info = {"All Files", (const char*)ICON_FA_FILE};
+        }
+      }
+
+      ImGui::TableNextRow();
+      ImGui::TableNextColumn();
+
+      // Alternate row colors
+      ImGui::TableSetBgColor(
+          ImGuiTableBgTarget_RowBg0,
+          ImGui::ColorConvertFloat4ToU32(
+              row % 2 ? ImGui::GetStyleColorVec4(ImGuiCol_TableRowBgAlt)
+                      : ImGui::GetStyleColorVec4(ImGuiCol_TableRowBg)));
+
+      auto next_node = node + 1;
+
+      ImGui::PushID(node->name.c_str());
+      {
+        auto icon_color = extension_info.m_icon_color;
+        if (icon_color)
```

Comment:
Would this work?
```cpp
auto icon_color = extension_info.m_icon_color.value_or(/* default color */);
ImGui::PushStyleColor(ImGuiCol_Text, *con_color);
rsl::Defer(ImGui::PopStyleColor());
```

Reply:
It wouldn't for two reasons. Firstly value_or keeps me from implicitly defaulting to the current color scheme (I would have to initialize it with the color from the style instance), and secondly (and more importantly) rsl::Defer would keep me from popping the color at the correct time (before text, after icon).